static/src/javascripts/projects/common/modules/component.spec.js

```diff
@@ -0,0 +1,426 @@
+// @flow
+
+import fetchJSON_ from 'lib/fetch-json';
+
+import { Component } from './component';
+
+const fetchJSON: JestMockFn<*, *> = (fetchJSON_: any);
+
+const mockResponse = {
+    html: '<p>html</p>',
+    other: '<p>other</p>',
+};
+
+const createComponent = (props?: Object = {}): Component => {
+    const component = new Component();
+    const defaults = {
+        endpoint: 'whatever',
+    };
+
+    Object.assign(component, defaults, props);
+
+    return component;
+};
+
+jest.mock('lib/fetch-json', () => jest.fn(() => Promise.resolve(mockResponse)));
+
+describe('Component', () => {
+    describe('fetch()', () => {
+        test('call fetched() with an endpoint', () => {
+            const component = createComponent({
+                fetched: jest.fn(),
+            });
+
+            return component.fetch().then(() => {
+                expect(component.fetched).toHaveBeenCalledWith(mockResponse);
+            });
+        });
+
+        test('not call fetched() without an endpoint', () => {
+            const component = createComponent({
+                endpoint: false,
+                fetched: jest.fn(),
+            });
+
+            return component.fetch().then(() => {
+                expect(component.fetched).not.toHaveBeenCalled();
+            });
+        });
+
+        test('extract the content of `html` in response by default', () => {
+            const component = createComponent({
+                fetched: jest.fn(),
+            });
+
+            return component.fetch().then(() => {
+                if (!component.elem) {
+                    return Promise.reject(
+                        new Error('.elem property should exist')
+                    );
+                }
+
+                if (component.elem) {
+                    expect(component.elem.tagName).toBe('P');
+                }
+
+                if (component.elem) {
+                    expect(component.elem.innerHTML).toBe('html');
+                }
```

Comment:
Could probably combine these `if` statements?

Reply:
Nope 😢 `expect()` could have sideeffects and therefore flow ... you know.